atc/gc/pipeline_collector.go

```diff
@@ -0,0 +1,88 @@
+package gc
+
+import (
+	"code.cloudfoundry.org/lager"
+	"context"
+	"time"
+
+	"code.cloudfoundry.org/lager/lagerctx"
+
+	"github.com/concourse/concourse/atc/db"
+)
+
+
+// pipelineCollector takes a parameter `ttl`, and checks all pipelines. If a
+// non-paused has no any build within `ttl`, then it will be paused; if a paused
+// pipeline has no any build within 2*`ttl`, then it will be destroyed. It `ttl`
+// is 0, then pipelineCollector will not touch any pipeline.
+type pipelineCollector struct {
+	pipelineFactory db.PipelineFactory
+	ttl             time.Duration
+}
+
+func NewPipelineCollector(pipelineFactory db.PipelineFactory, ttl time.Duration) *pipelineCollector {
+	return &pipelineCollector{
+		pipelineFactory: pipelineFactory,
+		ttl:             ttl,
+	}
+}
+
+func (p *pipelineCollector) Run(ctx context.Context) error {
+	logger := lagerctx.FromContext(ctx).Session("pipeline-collector")
+
+	logger.Debug("start")
+	defer logger.Debug("done")
+
+	if p.ttl == 0 {
+		logger.Debug("ttl-is-zero")
+		return nil
+	}
+
+	pipelines, err := p.pipelineFactory.AllPipelines()
+	if err != nil {
+		logger.Error("failed-to-get-pipelines", err)
+		return err
+	}
+	logger.Debug("after-get-all-pipelines", lager.Data{"total-pipeline-count": len(pipelines)})
+
+	for _, pipeline := range pipelines {
+		log := logger.Session("tick").WithData(lager.Data{"pipeline-id": pipeline.ID()})
+		if pipeline.Paused() {
```

Comment:
:thought_balloon: This could make it pretty easy to accidentally destroy your prod pipeline if you pause it while investigating an external issue that takes longer than expected to resolve. Paused pipelines will not produce any builds so they'd be subject to this garbage collector after enough time elapses.

What value do you think y'all would set for your own environment? I would imagine you'd need a pretty high TTL for pipelines to survive the weekend, too.

Reply:
We would not use this feature on production cluster. From my current thinking, for playground (experiment) cluster, TTL could be 7 days; and for staging clusters, TTL could be 30 days.